### PR TITLE
Hotfix: filter out unsatisfied nodes when calling AddPod in PodTopologySpread

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind bug
/sig scheduling

#### What this PR does / why we need it:
Filter out unsatisfied nodes when running the set of configured filter plugins 
for nominated pod on the given node, this will help to prevent potential scheduler crash.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/110726

#### Does this PR introduce a user-facing change?

```release-note
Fixed potential scheduler crash when scheduling with unsatisfied nodes in PodTopologySpread.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 1]
```diff
@@ -127,6 +127,12 @@ func (s *preFilterState) updateWithPod(updatedPod, preemptorPod *v1.Pod, node *v
 		return
 	}
 
+	requiredSchedulingTerm := nodeaffinity.GetRequiredNodeAffinity(preemptorPod)
+	match, _ := requiredSchedulingTerm.Match(node)
+	if !match {
+		return
+	}
```
Comment (damemi) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:134]: nit: `if match, _ := requiredSchedulingTerm.Match(node); !match {`

**Review by damemi — COMMENTED** [review]
/approve

**Comment by ahg-g** [comment]
looks good to me

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go (hunk 1 above)
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:134]: Thanks Mike, done :)

**Comment by kerthcet** [comment]
Updated and I'll cherry-pick this to other releases.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 2]
```diff
@@ -127,6 +127,11 @@ func (s *preFilterState) updateWithPod(updatedPod, preemptorPod *v1.Pod, node *v
 		return
 	}
 
+	requiredSchedulingTerm := nodeaffinity.GetRequiredNodeAffinity(preemptorPod)
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:130]: Add a comment:

```
// We only need to update the counters for the nodes that match the topology spreading policy, in other words, the node affinity.
```

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 3]
```diff
@@ -135,6 +140,12 @@ func (s *preFilterState) updateWithPod(updatedPod, preemptorPod *v1.Pod, node *v
 
 		k, v := constraint.TopologyKey, node.Labels[constraint.TopologyKey]
 		pair := topologyPair{key: k, value: v}
+
+		if s.TpPairToMatchNum[pair] == nil {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:144]: Let's remove this

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 4]
```diff
@@ -818,6 +819,31 @@ func TestPreFilterStateAddPod(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "add a pod with NotIn nodeAffinity",
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:823]:
```suggestion
			name: "add a pod that doesn't match the node affinity",
```

**Comment by alculquicondor** [comment]
quite complicated catch

thanks!

/approve

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go (hunk 2 above)
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:130]: done

**Comment by kerthcet** [comment]
Updated and squashed @alculquicondor

**Comment by ahg-g** [comment]
@kubernetes/release-managers 

/lgtm
/approve

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 5]
```diff
@@ -818,6 +819,31 @@ func TestPreFilterStateAddPod(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "add a pod that doesn't match the node affinity",
+			preemptor: st.MakePod().Name("p").Label("foo", "").NodeAffinityNotIn("foo", []string{"bar"}).
+				SpreadConstraint(1, "zone", v1.DoNotSchedule, st.MakeLabelSelector().Exists("foo").Obj()).
+				Obj(),
+			nodeIdx:  0,
+			addedPod: st.MakePod().Name("p-a1").Node("node-b").Label("foo", "").Label("zone", "zone2").Obj(),
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:828]: nit: p-a1 -> p-b1, zone:zone label is not needed

```suggestion
			addedPod: st.MakePod().Name("p-b1").Node("node-b").Label("foo", "").Obj(),
```

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 6]
```diff
@@ -818,6 +819,31 @@ func TestPreFilterStateAddPod(t *testing.T) {
...
+			existingPods: []*v1.Pod{
+				st.MakePod().Name("p-b1").Node("node-a").Label("foo", "").Label("zone", "zone1").Obj(),
+				st.MakePod().Name("p-b2").Node("node-b").Label("foo", "").Label("zone", "zone2").Obj(),
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:831]:
```suggestion
				st.MakePod().Name("p-b2").Node("node-b").Label("foo", "").Obj(),
```

**Review by Huang-Wei — COMMENTED** [review]
Thanks for the fix @kerthcet !

Some nits for the test.

**Comment by Huang-Wei** [comment]
/lgtm

@Verolop Good to go now.

**Pull request merged**